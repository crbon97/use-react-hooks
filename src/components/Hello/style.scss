.hello-container {
  width: 50%;
  margin: 0 auto;

  .greet {
    font-size: 1.5em;
    color: coral;
  }

  .hello-form {
    color: brown;

    .field-text {
      height: 2.5em;
      width: 50%;
      -webkit-border-radius: 5px;
      -moz-border-radius: 5px;
      border-radius: 5px;
      padding: 0 10px;
      border: 1px solid burlywood;
      margin-right: 10px;

      &:focus {
        -webkit-border-radius: 5px;
        -moz-border-radius: 5px;
        border-radius: 5px;
        box-shadow: 0 0 5px orange;
        border-color: orange;
        outline: none;
      }
    }

    .button {
      display: inline-block;
      height: 2.5em;
      width: 20%;
      border: 1px solid burlywood;
      background: transparent;
      border-radius: 8px;
      font-weight: bold;
      text-transform: uppercase;
      cursor: pointer;

      &:focus {
        outline: none;
        border-color: orange;
        box-shadow: 0 0 5px orange;
      }
    }
  }

}