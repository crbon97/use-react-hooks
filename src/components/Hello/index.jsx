import "./style.scss";

import React, { useRef, useState } from "react";

const defaultName = "guest";
const Hello = () => {
  const nameRef = useRef();
  const [name, setName] = useState(defaultName);
  const hanldeSubmit = () => {
    const inputedName = nameRef.current.value;
    if (!!inputedName) {
      setName(inputedName);
    } else {
      alert("Please input your name, thanks!");
    }
  };

  const formatedName = name
    .toLowerCase()
    .replace(/(\b[a-z]\s?)/g, word => word.toUpperCase());

  return (
    <div className="hello-container">
      {/* Rerence: https://www.fullstackreact.com/articles/an-introduction-to-hooks-in-react/#useref-hook-example */}
      <p className="greet" data-testid="greet">
        Hello, {formatedName}!
      </p>
      <div className="hello-form">
        <input
          ref={nameRef}
          type="text"
          className="field-text"
          placeholder="Please input your name!"
          data-testid="name"
        />
        <button
          className="button"
          type="button"
          onClick={hanldeSubmit}
          data-testid="submit"
        >
          Submit
        </button>
      </div>
    </div>
  );
};

export default Hello;
