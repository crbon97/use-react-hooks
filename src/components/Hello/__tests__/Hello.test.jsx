import { cleanup, fireEvent, getByTestId, render } from "react-testing-library";

import Hello from "..";
import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

describe("Test Hello component with `react-test-renderer`", () => {
  test("should render without error", () => {
    const hello = renderer.create(<Hello />);
    let tree = hello.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Test Hello component with `enzyme`", () => {
  test('should render "Hello, Guest!" when having no name inputed', () => {
    const hello = shallow(<Hello />);
    const expected_greet = "Hello, Guest!";
    const greet = hello.find(".greet");
    expect(greet.text()).toEqual(expected_greet);
  });

  // Can't use Enzyme to test React Hooks (useRef, ....)
  // Use "react-testing-library" instead
  /*
  test('should render "Hello, Phat!" when having inputed name', () => {
    const hello = shallow(<Hello />);
    const expected_greet = "Hello, Phat!";
    const greet = hello.find(".greet");

    // Set value = 'Phat' to input
    hello.simulate('change', {target: {value: 'phat'}})
    expect(greet.text()).toEqual(expected_greet);
  });
  */
});

describe("Test Hello component with `react-testing-library`", () => {
  beforeEach(cleanup);
  test('should render "Hello, Phat!" when having inputed name', () => {
    const { container } = render(<Hello />);
    const greetElement = getByTestId(container, "greet");
    const inputName = getByTestId(container, "name");
    const submitButton = getByTestId(container, "submit");

    fireEvent.change(inputName, { target: { value: "phat" } });
    fireEvent.click(submitButton);

    const expectedGreet = "Hello, Phat!";
    expect(greetElement.textContent).toEqual(expectedGreet);
  });
});
