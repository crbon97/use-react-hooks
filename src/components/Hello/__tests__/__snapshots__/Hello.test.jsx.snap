// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Hello component with \`react-test-renderer\` should render without error 1`] = `
<div
  className="hello-container"
>
  <p
    className="greet"
    data-testid="greet"
  >
    Hello, 
    Guest
    !
  </p>
  <div
    className="hello-form"
  >
    <input
      className="field-text"
      data-testid="name"
      placeholder="Please input your name!"
      type="text"
    />
    <button
      className="button"
      data-testid="submit"
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </div>
</div>
`;
